Understanding c++ templates (#1777)
section-engineering-education/engineering-education: articles/understanding-c++-templates/index.md

> ### Introduction

Reviewer: ```suggestion
## Introduction
```

A template is a way of writing a single function or class that can be reused by a group of similar functions or classes. In plain terminology, you can use templates to declare a single function or class that works with different data types. This article will take a look at templates and how to use them for generic programming in c++.

> ### Prerequisites

Reviewer: ```suggestion
## Prerequisites
```

To follow this article, you’ll need to have:
- A [Codeblocks IDE](http://www.codeblocks.org/downloads) to run the code.
- An understanding of the C++ language.

We can use the concept of templates in two different ways:
- `Class templates`
- `Function templates`

> ### How do templates work?

Reviewer: ```suggestion
## How do templates work?
```

Templates expand at compile time like macros. The difference is that the compiler does a type check before expanding the template. The principle is basic, source code only contains function/class, but many copies of the same function/class can be included in compiled code.

> ### Class templates

Reviewer: ```suggestion
## Class templates
```

A class template is a template that enables the developer to operate with generic data types. This enables the class to be used on several different data types as per the specifications without the need for each type to be re-written.
Quite often, you'll need a class implementation that is the same for all classes except for the data types used. For each data type, you'd have to declare a separate class or build many member variables and functions within a single class.

> This will bloat the code base and will be difficult to manage because of a change in one class that will need change on all classes/functions. But, class templates allow you to reuse the same code for all data types. Like normal arguments, we can transfer more than one type of data to templates as arguments if we need to.

Reviewer: ```suggestion
The declaration of separate classes that do the same thing for different data types will bloat the code base and make it difficult to manage. Any changes made in one class will need to be reflected in all the similar classes. 

Class templates allow you to reuse the same code for all data types. Like normal arguments, we can transfer more than one type of data to templates as arguments if we need to.
```

> #### Declaration of a class template

Reviewer: ```suggestion
### Declaration of a class template
```

Syntax:
```c++
template <class T>
class class_name
{
   ... .. ...
public:
   T var;
   T functionName(T arg);
   ... .. ...
};
```
Let’s break down the syntax as shown below:

- `template <class T>`- It is the keyword to declare a template.
- `class` - It is a user-defined type.
- `class_name` - It is the class name for the class.
- `T` - This is the argument for the template that is a placeholder for the data type which will be used.
- `var` - It is a variable of the class of the generic data type.
- ` T functionName(T arg)` - It is a function of the generic data type.

> #### How to create a class template object?

Reviewer: ```suggestion
### How to create a class template object?
```

To create the function declaration of `T larger()` outside the class, we first declare the generic data type `template <class T>`. Note that if you create the function declaration outside the class without it, the compiler will generate errors. 

To build the function declaration outside the class template, we first declare the generic return type `T`  and then the class name `sample`. Right after the class name, we add `<T>` to tell the compiler that the function’s template  argument is the same one that we are using for the class.

We compare the two digits within the function definition. We have created a class object using the same format we saw before the program:

`  sample <int> big(120, 80);`

Every time we are using the class template, it is required to tell it what data type we are substituting for `T`. In this instance, we are substituting `T` for `int`. Afterward, we created the object `big` and passed two integer numbers. We finish the program by calling the function `T larger()`.

> #### Advantages of class templates

Reviewer: ```suggestion
### Advantages of class templates
```

The main advantages of using a class template are:
- You need to define only one class that will operate with various types of data.
- Only those data types for which the template has been used in the application produce instances of this template class at compile time.
- As we have already established, the creation and use of a template class will cut the development efforts and lines of code. It will also decrease the difficulty and time of debugging the software for any problems because you only deal with 1 class.

> ### Function templates

Reviewer: ```suggestion
## Function templates
```

A [function template](https://bcastudyguide.wordpress.com/unit-4-generic-function) acts similarly to a normal function, with one key difference. A single function template can [interact with different data types at once](https://bcastudyguide.wordpress.com/unit-4-generic-function), but with a single normal function, only one set of data types can work.

Examples of OOP's polymorphism feature are both function overloading and function templates.  If you need to perform similar operations for two or more data types, you use overloading functions to create two functions. When several functions perform identical operations, templates are used. Using function templates, but, would be a safer method since you can do the same task with less and more maintainable code

> #### Declaration of a function template?

Reviewer: ```suggestion
### Declaration of a function template?
```

The Output will be:
```bash
Enter two integer numbers:
18 10
18 is larger.
Enter two floating-point numbers:
4.5 9.3
9.3 is larger.
Enter two characters:
a m
m has a larger ASCII value.
```
Our program above takes user input thus for us to view the output, we can use the following values:

Integer numbers: 18, 10
Floating-point numbers: 4.5, 9.3
Characters: a, m

The function template `Large()` is declared in the above program, which accepts two `d1` and `d2` arguments of data type `T`. `T` indicates any type can be an argument. The larger of the two arguments is returned by the `Large() function` using a conditional operator.

Variables of three distinct data types are declared within the `main()` function. The variables will then be passed as normal functions to the `Large()` function prototype. When we pass an integer to the function template during runtime, the compiler knows that the `Large()` function must accept `int` arguments.

Likewise, when `float` and `char` data are passed, the `Large()` function is generated because it recognizes the argument data types. This way, three identical normal functions were replaced by using a single function prototype.

> #### Variadic function template

Reviewer: ```suggestion
### Variadic function template
```

`display(1, 2, 3.14, “You can pass me any number of arguments”, “I will display\n”);` is evaluated in the following way:

The compiler will first resolve the statement into:

```c++
cout<< 1 <<endl ;
display(2, 3.14, "You can pass me any number of arguments", 
      "I will display\n");
```
Next, the compiler looks for a `display()` function that can take those arguments and, as a result, the function is executed again in the same way:

```c++
cout<< 2 <<endl ;
display(3.14, "You can pass me any number of arguments", 
      "I will display\n");
```
It is resolved into the following forms once more:

```c++
cout<< 3.14 <<endl ;
display("You can pass me any number of arguments", 
      "I will display\n");
```

```c++
cout<< "You can pass me any number of arguments" <<endl ;
display("I will display\n");
```

```c++
cout<< "I will display\n" <<endl ;
display();
```
The compiler at this point searches for a function overload that fits the empty function, that is, a function that takes no arguments. All functions with one or more arguments are matched to the variadic function, while the empty are matched to the empty functions.

> ### Conclusion 

Reviewer: ```suggestion
## Conclusion 
```
